util/helm/client.go

```diff
@@ -98,57 +98,81 @@ func (c *nativeHelmChart) ExtractChart(chart string, version *semver.Version) (s
 	if err != nil {
 		return "", nil, err
 	}
-	chartPath := c.getChartPath(chart, version)
 
-	c.repoLock.Lock(chartPath)
-	defer c.repoLock.Unlock(chartPath)
+	// always use Helm V3 since we don't have chart content to determine correct Helm version
+	helmCmd, err := NewCmdWithVersion(c.repoPath, HelmV3, c.enableOci)
 
-	exists, err := fileExist(chartPath)
 	if err != nil {
 		return "", nil, err
 	}
-	if !exists {
-		// always use Helm V3 since we don't have chart content to determine correct Helm version
-		helmCmd, err := NewCmdWithVersion(c.repoPath, HelmV3, c.enableOci)
+	defer helmCmd.Close()
+
+	_, err = helmCmd.Init()
+	if err != nil {
+		return "", nil, err
+	}
+
+	// throw away temp directory that stores extracted chart and should be deleted as soon as no longer needed by returned closer
+	tempDir, err := ioutil.TempDir("", "helm")
+	if err != nil {
+		return "", nil, err
+	}
 
+	// OCI registry maintain own cache and client provide methods to extract chart into temp directories
+	if c.enableOci {
+		_, err = helmCmd.Login(c.repoURL, c.creds)
 		if err != nil {
 			return "", nil, err
 		}
-		defer helmCmd.Close()
 
-		_, err = helmCmd.Init()
+		defer func() {
+			_, _ = helmCmd.Logout(c.repoURL, c.creds)
+		}()
+
+		// 'helm chart pull' ensures that chart is downloaded into local repository cache
+		_, err = helmCmd.ChartPull(c.repoURL, chart, version.String())
 		if err != nil {
 			return "", nil, err
 		}
 
-		if c.enableOci {
-			_, err = helmCmd.Login(c.repoURL, c.creds)
-			if err != nil {
-				return "", nil, err
-			}
-			defer func() {
-				_, _ = helmCmd.Logout(c.repoURL, c.creds)
-			}()
+		// 'helm chart export' copies cached chart into temp directory
+		_, err = helmCmd.ChartExport(c.repoURL, chart, version.String(), tempDir)
+		if err != nil {
+			return "", nil, err
 		}
 
+		return path.Join(tempDir, normalizeChartName(chart)), io.NewCloser(func() error {
+			return os.RemoveAll(tempDir)
+		}), nil
+	}
+
+	// non OCI registry don't have effective caching system
+	// so we have to manually cache chart in tar file
+	chartPath := c.getChartPath(chart, version)
+
+	c.repoLock.Lock(chartPath)
+	defer c.repoLock.Unlock(chartPath)
+
+	// check if tar already downloaded
+	exists, err := fileExist(chartPath)
+	if err != nil {
+		return "", nil, err
+	}
+	if !exists {
+		// download chart using 'helm fetch' command and create tar to avoid re-downloading chart in future
+
```

Comment:
Great catch. Thank you!

I had a wrong assumption - OCI registry charts have to be cached in tar files same as non-OCI chart (we blow away helm cache including OCI registry cache). Rewrote code so that OCI charts are also stored in pre-defined tar file location that can be deleted by force refresh.